Implement undo_format_as fixes #1301

JesseLivezey opened the pull request:
This is motivated by the following problem:

FlattenerLayers interact with their raw_layers when costs are called from them and when monitoring channels are requested. Both of these things generally do not want the FlattenerLayer output, but rather the output of the layers in the CompositeLayer the FlattenerLayer wraps. This can't be done with another format_as since this adds more theano ops and then the costs and monitoring channels can't get the actual results of fprop easily.

Fixes problems referenced here:
https://github.com/lisa-lab/pylearn2/issues/1301

and might help:
https://github.com/lisa-lab/pylearn2/pull/1290

- [x] confirm that this fixes known problems (did this locally)
 - [x] add tests to pylearn2
- [x] write tests to confirm that this works for all relevant spaces
 - [x] write tests for symbolic undo
 - [x] write tests for numeric undo
 - [x] Check dtype conversion
- [x] remove duplicate code with _undo_cast, etc.

JesseLivezey commented:
@lamblin, this PR should help with some of the FlattenerLayer problems that have come up recently. I basically wrote ```space.undo_format_as()``` by undoing each symbolic op from ```space.format_as()```.

I wrote a bunch of tests to check direct space conversions. I can also add a test network that was broken before--a FlattenerLayer that wraps ConvLayers.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +548,129 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        return self._undo_format_as(is_numeric=True,
+                                    batch=batch,
+                                    space=space)
```
I'm not sure why delegate the work to _undo_format_as, since we could call np_format_as here directly.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +548,129 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        return self._undo_format_as(is_numeric=False,
+                                    batch=batch,
+                                    space=space)
```
Same thing here, what is done by _undo_format_as(is_numeric=False, ...) could simply be put here.

JesseLivezey commented:
@lamblin, I was mostly just following the structure of format_as for no good reason. I've removed the unnecessary methods.

I also cleaned up some of the duplicated code with an _undo_op function.

lamblin commented:
Makes a lot of sense to me.
Could you please rebase on the latest master, and remove the merge commit? If you use `git rebase -i`, the merge commit may automatically disappear, or you could explicitly remove it.
Then you would have to force the `git push` to this branch.

JesseLivezey commented:
@lamblin, I've rebased onto master, but there is a failure coming from a ```float < str``` comparison when the monitoring channels are being sorted (I think). It looks like it is only happening in the python 3 tests.

dwf commented:
Build failing. Is this still WIP? Can you add [WIP] to the title?

JesseLivezey commented:
@dwf, it seems to be failing because of a python 3 problem that I don't think I've caused. I described in in my previous comment.

lamblin commented:
Trying to trigger merge with the latest master, which may have fixed the python3 issue.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +576,87 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        Returns
+        -------
+        WRITEME
```
You can simply put the type here, for instance `numpy.ndarray, or one of the scipy.sparse matrices` instead of WRITEME. You do not have to name the returned value.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +576,87 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        Returns
+        -------
+        WRITEME
```
idem.

JesseLivezey replied:
Fixes these in new commit.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +576,87 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        Parameters
+        ----------
+        batch : atheano symbol, or a nested tuple thereof
```
" a theano"

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -548,6 +576,87 @@ def _format_as_impl(self, is_numeric, batch, target_space):
+
+        Returns
+        -------
+        The batch, converted from target_space, back to self.
```
First line has to be a type

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -1998,7 +2235,8 @@ def recursive_shared(batch):
     def _format_as_impl(self, is_numeric, batch, space):
         """
-        Supports formatting to a single VectorSpace, or to a CompositeSpace.
+        Supports undo formatting to a single VectorSpace,or to a
```
This change should probably be reverted, this is the docstring of `_format_as_impl`, not `_undo_format_as_impl`

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -2084,6 +2322,105 @@ def recursive_format_as(orig_space, batch, dest_space):
+    @functools.wraps(Space._undo_format_as_impl)
+    def _undo_format_as_impl(self, batch, space):
+        """
+        Supports formatting to a single VectorSpace, or to a CompositeSpace.
```
And the other way around here.

JesseLivezey replied:
Changed these back and make "undo" version more clear.

lamblin commented on pylearn2/space/__init__.py:
```diff
@@ -2321,4 +2658,3 @@ def _batch_size_impl(self, is_numeric, batch):
         # There is no way to know how many examples would actually
         # have been in the batch, since it is empty. We return 0.
         self._validate(is_numeric, batch)
-        return 0
```
Why remove that line? the comment says "We return 0."

JesseLivezey replied:
This wasn't intentional, fixed in new commit.

lamblin commented:
The buildbot is passing now.

JesseLivezey commented:
@lamblin, I've made the changes you've requested.

lamblin commented:
Great, thanks!
